# Task1

In [4]:
#pip uninstall googletrans
!pip install googletrans==3.1.0a0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.1/55.1 KB 3.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 27.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.6/42.6 KB 5.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.4/133.4 KB 18.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.0/65.0 KB 8.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.6/53.6 KB 6.8 MB/s eta 0:00:00
  Created wheel for googletrans: filename=googletrans-3.1.0a0-py3-none-any.whl size=16368 sha256=67e26ff4e83d8e8e2365e456cb63a6080dbe276d40b7396fd03dbb78e6ac6530
  Stored in directory: /root/.cache/pip/wheels/dd/59/af/8d6c96a719763990f1c548e36b17d9efdfb767f42f7ff39f53
Successfully built googletrans
  Attempting uninstall: chardet
    Found existing installation: chardet 4.0.0
 

In [5]:
import googletrans

In [6]:
from googletrans import Translator
import pandas as pd

In [7]:
df=pd.read_csv('/content/sample_data/English.csv')

In [8]:
df.head()

English words/sentences
0                     Hi.
1                    Run!
2                    Run!
3                    Who?
4                    Wow!

In [9]:
df['English words/sentences'][2]

'Run!'

In [10]:
df['processed']=df['English words/sentences'].str.replace("[^a-zA-Z0-9]"," ")

<ipython-input-10-30fb80eda9eb>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['processed']=df['English words/sentences'].str.replace("[^a-zA-Z0-9]"," ")


In [11]:
df.processed.unique()

array(['Hi ', 'Run ', 'Who ', ...,
       'Death is something that we re often discouraged to talk about or even think about  but I ve realized that preparing for death is one of the most empowering things you can do  Thinking about death clarifies your life ',
       'Since there are usually multiple websites on any given topic  I usually just click the back button when I arrive on any webpage that has pop up advertising  I just go to the next page found by Google and hope for something less irritating ',
       'If someone who doesn t know your background says that you sound like a native speaker  it means they probably noticed something about your speaking that made them realize you weren t a native speaker  In other words  you don t really sound like a native speaker '],
      dtype=object)

In [12]:
df.processed=[row.lower() for row in df.processed]

In [13]:
df.processed=[row.strip() for row in df['processed']]

In [14]:
df.head()

English words/sentences processed
0                     Hi.        hi
1                    Run!       run
2                    Run!       run
3                    Who?       who
4                    Wow!       wow

In [74]:
df.shape

(175621, 2)

In [17]:
df2=df[:100]

In [76]:
df2.head()

English words/sentences processed
0                     Hi.        hi
1                    Run!       run
2                    Run!       run
3                    Who?       who
4                    Wow!       wow

In [15]:
translator=Translator()

### Since it is taking ore time for translating all 175621 rows, iwill consider a small fraction of dataframe

### English to Spanish

In [18]:
df2['Spanish']=df2['processed'].apply(translator.translate,src='en',dest='es').apply(getattr,args=('text',))

<ipython-input-18-38d29fce7d06>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Spanish']=df2['processed'].apply(translator.translate,src='en',dest='es').apply(getattr,args=('text',))


In [19]:
df2.head()

English words/sentences processed Spanish
0                     Hi.        hi    Hola
1                    Run!       run  correr
2                    Run!       run  correr
3                    Who?       who     OMS
4                    Wow!       wow    Guau

### English to French

In [20]:
df2['French']=df2['processed'].apply(translator.translate,src='en',dest='fr').apply(getattr,args=('text',))

<ipython-input-20-83719bff4fb4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['French']=df2['processed'].apply(translator.translate,src='en',dest='fr').apply(getattr,args=('text',))


In [21]:
df2.head()

English words/sentences processed Spanish  French
0                     Hi.        hi    Hola   Salut
1                    Run!       run  correr  courir
2                    Run!       run  correr  courir
3                    Who?       who     OMS     OMS
4                    Wow!       wow    Guau    Ouah

### English to German

In [22]:
df2['german']=df2['processed'].apply(translator.translate,src='en',dest='de').apply(getattr,args=('text',))

<ipython-input-22-8af6e7cdfd82>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['german']=df2['processed'].apply(translator.translate,src='en',dest='de').apply(getattr,args=('text',))


In [23]:
df2.head()

English words/sentences processed Spanish  French  german
0                     Hi.        hi    Hola   Salut      Hi
1                    Run!       run  correr  courir  laufen
2                    Run!       run  correr  courir  laufen
3                    Who?       who     OMS     OMS     WHO
4                    Wow!       wow    Guau    Ouah     Wow

# Task2

In [83]:
pip install textblob

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [27]:
from textblob import Word

In [28]:
def correct_word_spelling(word):
  words=Word(word)
  result=words.correct()
  return result

In [29]:
correct_word_spelling('apppleee')

'apppleee'

In [30]:
from textblob import TextBlob

In [31]:
def sentence_correct_spelling(sentence):
  sentence=TextBlob(sentence)
  result=sentence.correct()
  return result

In [41]:
sentence_correct_spelling('A sentencee to checkk')

TextBlob("A sentence to check")

In [33]:
df = pd.read_table('/content/sample_data/spelling_word.txt',delimiter=':',usecols=[1], header=None,names=['mis-spelled'])

In [34]:
df.head()

mis-spelled
0           apeal
1          appple
2        encorage
3       perminant
4   mathematicaly

In [35]:
df['correct_spelling']=df['mis-spelled'].apply(lambda x:correct_word_spelling(x))

In [36]:
df.head()

mis-spelled correct_spelling
0           apeal           appeal
1          appple            apple
2        encorage        encourage
3       perminant        perminant
4   mathematicaly     mathematical

In [37]:
df2=pd.read_table('/content/sample_data/sentence_spelling.txt',names=['sentence_correction'])

In [38]:
df2.head()

sentence_correction
0   I really appreciat
1           I am sory.
2  How does tat sound?
3    That sounds grat.
4         neverr mind.

In [43]:
for x in df2['sentence_correction']:
  sentence=TextBlob(x)
  result=sentence.correct()
  print(result)

I really appreciate
I am story.
Now does tat sound?
That sounds great.
never mind.
I learn English.
I do not understand.
Would you repeat that please?
please take slower?
Thank you. That helps a lost.
I do not head a bag.
An you spell that?
help mi
give a different color
I do not no my size.
There can I find
so one help me
Is this delivered
You have compromised yourself seriously.
I was only Grown Prince then.
I was young.
I am thirty now
It must be recovered
He had tried
I failed.
It must be bought
The will not see
It is swollen
I will be with you in a moment.
wen is our meeting
Please cal me
open the later quickly.
The patted me on the shoulder.
The plays tennis after school.
Once upon a time
an elderly hero maid his home in a pond full of fish.
He was stir
he was slow in his old age
he didn't find it easy to catch his lunch.
However, he decided to use his wits
he went to see a crab
a great chaterbox
and in a mock frightened voice
told him the latest rumour.
Certain birds are my frie

In [45]:
df2.drop(['corrected_sentence'],axis=1,inplace=True)

# Task3

In [70]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import re
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from scipy.sparse import hstack
from sklearn.multiclass import OneVsRestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

In [51]:
df=pd.read_csv('/content/sample_data/UpdatedResumeDataSet.csv',encoding='utf-8')

In [53]:
df.head()

Category                                             Resume
0  Data Science  Skills * Programming Languages: Python (pandas...
1  Data Science  Education Details \r\nMay 2013 to May 2017 B.E...
2  Data Science  Areas of Interest Deep Learning, Control Syste...
3  Data Science  Skills • R • Python • SAP HANA • Table...
4  Data Science  Education Details \r\n MCA   YMCAUST,  Faridab...

In [55]:
df.shape

(962, 2)

In [59]:
df['Category'].nunique()

25

In [57]:
df['Category'].value_counts()

Java Developer               84
Testing                      70
DevOps Engineer              55
Python Developer             48
Web Designing                45
HR                           44
Hadoop                       42
Blockchain                   40
ETL Developer                40
Operations Manager           40
Data Science                 40
Sales                        40
Mechanical Engineer          40
Arts                         36
Database                     33
Electrical Engineering       30
Health and fitness           30
PMO                          30
Business Analyst             28
DotNet Developer             28
Automation Testing           26
Network Security Engineer    25
SAP Developer                24
Civil Engineer               24
Advocate                     20
Name: Category, dtype: int64

In [63]:
df['Resume'][0]

'Skills * Programming Languages: Python (pandas, numpy, scipy, scikit-learn, matplotlib), Sql, Java, JavaScript/JQuery. * Machine learning: Regression, SVM, Naïve Bayes, KNN, Random Forest, Decision Trees, Boosting techniques, Cluster Analysis, Word Embedding, Sentiment Analysis, Natural Language processing, Dimensionality reduction, Topic Modelling (LDA, NMF), PCA & Neural Nets. * Database Visualizations: Mysql, SqlServer, Cassandra, Hbase, ElasticSearch D3.js, DC.js, Plotly, kibana, matplotlib, ggplot, Tableau. * Others: Regular Expression, HTML, CSS, Angular 6, Logstash, Kafka, Python Flask, Git, Docker, computer vision - Open CV and understanding of Deep learning.Education Details \r\n\r\nData Science Assurance Associate \r\n\r\nData Science Assurance Associate - Ernst & Young LLP\r\nSkill Details \r\nJAVASCRIPT- Exprience - 24 months\r\njQuery- Exprience - 24 months\r\nPython- Exprience - 24 monthsCompany Details \r\ncompany - Ernst & Young LLP\r\ndescription - Fraud Investigatio

In [64]:
def cleanResume(resumeText):
    resumeText = re.sub('httpS+s*', ' ', resumeText)  # remove URLs
    resumeText = re.sub('RT|cc', ' ', resumeText)  # remove RT and cc
    resumeText = re.sub('#S+', '', resumeText)  # remove hashtags
    resumeText = re.sub('@S+', '  ', resumeText)  # remove mentions
    resumeText = re.sub('[%s]' % re.escape("""!"#$%&'()*+,-./:;<=>?@[]^_`{|}~"""), ' ', resumeText)  # remove punctuations
    resumeText = re.sub(r'[^x00-x7f]',r' ', resumeText) 
    resumeText = re.sub('s+', ' ', resumeText)  # remove extra whitespace
    return resumeText

In [67]:
df['cleaned_resume'] = df.Resume.apply(lambda x: cleanResume(x))
df.head()

Category                                             Resume  \
0  Data Science  Skills * Programming Languages: Python (pandas...   
1  Data Science  Education Details \r\nMay 2013 to May 2017 B.E...   
2  Data Science  Areas of Interest Deep Learning, Control Syste...   
3  Data Science  Skills • R • Python • SAP HANA • Table...   
4  Data Science  Education Details \r\n MCA   YMCAUST,  Faridab...   

                                      cleaned_resume  
0  Skill    Programming Language   P thon  panda ...  
1  Education Detail    Ma  2013 to Ma  2017 B E  ...  
2  Area  of Intere t Deep Learning  Control S  te...  
3  Skill      R     P thon     SAP HANA     Table...  
4  Education Detail     MCA   YMCAUST   Faridabad...

In [69]:
var_mod = ['Category']
le = LabelEncoder()
for i in var_mod:
    df[i] = le.fit_transform(df[i])
df.head()

Category                                             Resume  \
0         6  Skills * Programming Languages: Python (pandas...   
1         6  Education Details \r\nMay 2013 to May 2017 B.E...   
2         6  Areas of Interest Deep Learning, Control Syste...   
3         6  Skills • R • Python • SAP HANA • Table...   
4         6  Education Details \r\n MCA   YMCAUST,  Faridab...   

                                      cleaned_resume  
0  Skill    Programming Language   P thon  panda ...  
1  Education Detail    Ma  2013 to Ma  2017 B E  ...  
2  Area  of Intere t Deep Learning  Control S  te...  
3  Skill      R     P thon     SAP HANA     Table...  
4  Education Detail     MCA   YMCAUST   Faridabad...

In [71]:
#Data Preprocessing
requiredText = df['cleaned_resume'].values
requiredTarget = df['Category'].values
word_vectorizer = TfidfVectorizer(sublinear_tf=True, stop_words='english', max_features=1500)
word_vectorizer.fit(requiredText)
WordFeatures = word_vectorizer.transform(requiredText)

In [72]:
X_train,X_test,y_train,y_test = train_test_split(WordFeatures,requiredTarget,random_state=0, test_size=0.2)
print(X_train.shape)
print(X_test.shape)
clf = OneVsRestClassifier(KNeighborsClassifier())
clf.fit(X_train, y_train)
prediction = clf.predict(X_test)

(769, 1500)
(193, 1500)


In [73]:
print('Accuracy of KNeighbors Classifier on training set: {:.2f}'.format(clf.score(X_train, y_train)))
print('Accuracy of KNeighbors Classifier on test set: {:.2f}'.format(clf.score(X_test, y_test)))
print("n Classification report for classifier %s:n%sn" % (clf, metrics.classification_report(y_test, prediction)))

Accuracy of KNeighbors Classifier on training set: 0.99
Accuracy of KNeighbors Classifier on test set: 0.99
n Classification report for classifier OneVsRestClassifier(estimator=KNeighborsClassifier()):n              precision    recall  f1-score   support

           0       1.00      1.00      1.00         3
           1       1.00      1.00      1.00         3
           2       1.00      0.80      0.89         5
           3       1.00      1.00      1.00         9
           4       1.00      1.00      1.00         6
           5       1.00      1.00      1.00         5
           6       1.00      1.00      1.00         9
           7       1.00      1.00      1.00         7
           8       1.00      0.91      0.95        11
           9       1.00      1.00      1.00         9
          10       1.00      1.00      1.00         8
          11       0.90      1.00      0.95         9
          12       1.00      1.00      1.00         5
          13       1.00      1.00      1.